test/e2e/test/step.go

```diff
@@ -32,23 +45,147 @@ func (l StepList) WithStep(testStep Step) StepList {
+				if err != nil {
+					logf.Log.Error(err, "while deleting elastic resources")
+				}
+			}()
+			// Only run eck diagnostics after each run when job is
+			// run from CI, which provides a 'job-name' flag.
+			if ctx.JobName != "" {
+				once.Do(initGSUtil)
+				logf.Log.Info("running eck-diagnostics job")
+				runECKDiagnostics(ctx, t.Name(), ts)
+				uploadDiagnosticsArtifacts()
+			}
+		}
+	}
+}
+
+func initGSUtil() {
+	cmd := exec.Command("gcloud", "auth", "activate-service-account", "--key-file=/etc/gcp/credentials.json")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Env = ensureTmpHomeEnv(cmd.Environ())
+	if err := cmd.Run(); err != nil {
+		log.Error(err, fmt.Sprintf("while initializing gsutil: %s", err))
+	}
+}
+
+func runECKDiagnostics(ctx Context, testName string, step Step) {
+	otherNS := append([]string{ctx.E2ENamespace}, ctx.Operator.ManagedNamespaces...)
+	// The following appends the test, and it's sub-test names together with a '-'.
+	// Example: For TestAutoscalingLegacy/Secrets_should_eventually_be_created:
+	// testName: TestAutoscalingLegacy, step.Name: Secrets_should_eventually_be_created
+	fullTestName := fmt.Sprintf("%s-%s", testName, step.Name)
+	// Convert any spaces to "_", and "/" to "-" in the test name.
+	normalizedTestName := strings.ReplaceAll(strings.ReplaceAll(fullTestName, " ", "_"), "/", "-")
+	cmd := exec.Command("eck-diagnostics", "--output-directory", "/tmp", "-n", fmt.Sprintf("eck-diagnostic-%s.zip", normalizedTestName), "-o", ctx.Operator.Namespace, "-r", strings.Join(otherNS, ","), "--run-agent-diagnostics") //nolint:gosec
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Env = ensureTmpHomeEnv(cmd.Environ())
+	if err := cmd.Run(); err != nil {
+		log.Error(err, fmt.Sprintf("while running eck-diagnostics: %s", err))
+	}
+}
+
+func uploadDiagnosticsArtifacts() {
+	ctx := Ctx()
+	cmd := exec.Command("gsutil", "cp", "/tmp/*.zip", fmt.Sprintf("gs://eck-e2e-buildkite-artifacts/jobs/%s/%s/", ctx.JobName, ctx.BuildNumber)) //nolint:gosec
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Env = ensureTmpHomeEnv(cmd.Environ())
+	if err := cmd.Run(); err != nil {
+		log.Error(err, fmt.Sprintf("while running gsutil: %s", err))
+	}
+}
+
+func ensureTmpHomeEnv(env []string) []string {
+	found := false
+	for i, e := range env {
+		if strings.Contains(e, "HOME=") {
+			env[i] = "HOME=/tmp"
+			found = true
+		}
+	}
+	if !found {
+		env = append(env, "HOME=/tmp")
+	}
+	return env
+}
+
+// This simulates "kubectl delete elastic" in the e2e namespace.
+func deleteElasticResources() error {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		log.Error(err, "while getting in cluster config")
+		return err
+	}
+	clntset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		log.Error(err, "while getting clientset")
+		return err
+	}
+
+	type version string
+	groupVersionMap := map[string][]version{}
+
+	apiGroups, _, _ := clntset.Discovery().ServerGroupsAndResources()
```

Comment:
Error ignored. Also I think the second return  argument already contains all the information needed for the deletion and makes the `NewDiscoveryCategoryExpander` unnecessary. It is a list of all API resources for each API Group/Version each entry contains all resources and sub-resources for each version including the category.

Reply:
I'll work through testing this and update with more information.